Fix the tests that assume 'Europe/Saratov' as invalid timezone

mmokhi opened the pull request:
Some timezone-related tests assume "Europe/Saratov" as a deprecated timezone.
While this seems to be actually true according to [unicode.org's reference](https://github.com/unicode-org/icu/blob/master/icu4c/source/data/misc/metaZones.txt#L3890) but
PHP [seems](https://www.php.net/manual/en/timezones.europe.php) keeping Saratov and bunch of others as valid locations,
which makes false-failures on Symfony's tests.
This patch fixes those failures.

Sponsored-by: Platform.sh

| Q             | A
| ------------- | ---
| Branch?       | master for features / 3.4, 4.2 or 4.3 for bug fixes <!-- see below -->
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| BC breaks?    | no     <!-- see https://symfony.com/bc -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tests pass?   | yes    <!-- please add some, will be required by reviewers -->
| Fixed tickets | #...   <!-- #-prefixed issue number(s), if any -->
| License       | MIT
| Doc PR        | symfony/symfony-docs#... <!-- required for new features -->

<!--
Replace this notice by a short README for your feature/bugfix. This will help people
understand your PR and can be used as a start for the documentation.

Additionally (see https://symfony.com/roadmap):
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too).
 - Features and deprecations must be submitted against the master branch.
-->

fabpot commented:
/cc @ro0NL

ro0NL commented:
@mmokhi  the tests ensure we exclude **valid** timezone IDs, that cannot be used with `\IntlTimeZone`. See https://3v4l.org/fQ27j

`Europe/Unknown` is never a valid timezone ID, so we need to find an older deprecated timezone, or perhaps skip the tests if the intl version is incompatible.

nicolas-grekas commented:
> skip the tests if the intl version is incompatible

I think that's the issue then, right? How do we do this?

mmokhi commented:
> @mmokhi the tests ensure we exclude **valid** timezone IDs, that cannot be used with `\IntlTimeZone`. See https://3v4l.org/fQ27j
> 

Is there actually a **valid** timezone that isn't usable with icu :thinking: ? 
Then why wouldn't icu add it to the list?

> `Europe/Unknown` is never a valid timezone ID, so we need to find an older deprecated timezone, or perhaps skip the tests if the intl version is incompatible.

looking at the links I posted above, I see deprecated names like _Bering_ and _Dutch_Guiana_  
But I wonder if they are being considered **valid**, since they are deprecated, why would they be treated different than something like _Europe/Unknown_

So yep, I guess either we can use them if you think it'll do the work, or better just skipping the test until we find such timezones :thinking: :slightly_smiling_face:

Commit:
Fix the tests that assume 'Europe/Saratov' as invalid timezone

Some timezone-related tests assume "Europe/Saratov" as a deprecated timezone.
While this seems to be actually true according to unicode.org's reference[1],
but PHP seems keeping Saratov[2] and bunch of others as valid locations,
which makes false-failures on Symfony's tests.
This patch fixes these issues.

[1] https://github.com/unicode-org/icu/blob/master/icu4c/source/data/misc/metaZones.txt#L3890
[2] https://www.php.net/manual/en/timezones.europe.php

Sponsored-by: Platform.sh

ro0NL commented:
> I wonder if there's actually a valid timezone that isn't usable with \IntlTimeZone thinking ?

Yes, `Europe/Saratov` :sweat_smile: but not on all machines apparently :thinking: 

> Then why wouldn't icu add it to the list?

They do;

https://github.com/symfony/symfony/blob/9cc2a4ecd1e6798e459c876bf2ea11bb94ebfb00/src/Symfony/Component/Intl/Resources/data/timezones/en.json#L368

The issue is purely a detail of `IntlTimeZone` which doesnt seem to handle deprecated/expired timezones and in this case uses `Etc/Unknown` instead. The same TZ does work with native `\DateTimeZone`

> Maybe the online service doesn't have intl module loaded in their .ini file?

It does https://3v4l.org/fgD9C, i don't understand what's causing the difference at this point :thinking: 

Can you try to find a timezon on your machine that doesnt work `IntlTimeZone`, e.g. like europe/saratov isnt working in my case.

ro0NL commented:
Im using https://hub.docker.com/r/jakzal/php-intl and even on `jakzal/php-intl:7.2-49.1.2` this TZ is already unavailable 

```
$ docker run -it jakzal/php-intl:7.2-49.1.2 bash
root@2f5987922e7c:/# php -r "var_dump('Etc/Unknown' === IntlTimeZone::createTimeZone('Europe/Saratov')->getID(), INTL_ICU_VERSION);"
bool(true)
string(6) "49.1.2"
```

mmokhi commented:
> The issue is purely a detail of `IntlTimeZone` which doesnt seem to handle deprecated/expired timezones and in this case uses `Etc/Unknown` instead. The same TZ does work with native `\DateTimeZone`
>

Yep, I believe so as well :slightly_smiling_face: 
However, I'm still confused why does it work for you :smile:  because [official PHP reference](https://www.php.net/manual/en/timezones.europe.php) maintains the `Europe/Saratov` as a valid point :thinking:  Or am I missing something?
 
> > Maybe the online service doesn't have intl module loaded in their .ini file?
> 
> It does https://3v4l.org/fgD9C, i don't understand what's causing the difference at this point thinking
> 
> Can you try to find a timezon on your machine that doesnt work `IntlTimeZone`, e.g. like europe/saratov isnt working in my case.

I see `America/Bering` (used to be valid timezone, not anymore. and doesn't exist in PHP-refs as well). is it working for you?

> Im using https://hub.docker.com/r/jakzal/php-intl and even on `jakzal/php-intl:7.2-49.1.2` this TZ is already unavailable
> 
> ```
> $ docker run -it jakzal/php-intl:7.2-49.1.2 bash
> root@2f5987922e7c:/# php -r "var_dump('Etc/Unknown' === IntlTimeZone::createTimeZone('Europe/Saratov')->getID(), INTL_ICU_VERSION);"
> bool(true)
> string(6) "49.1.2"
> ```

if it helps here's my PHP version result:

```
PHP 7.2.17 (cli) ( ZTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.2.17, Copyright (c) 1999-2018, by Zend Technologies
    with Xdebug v2.6.0, Copyright (c) 2002-2018, by Derick Rethans
```

ro0NL commented:
i think the issue is PHP (`\DateTimeZone`) uses an internal version of the timezone DB (https://www.iana.org/time-zones)

https://www.php.net/manual/en/timezones.php seems to confirm

> Note: The latest version of the timezone database can be installed via PECL's » timezonedb.
> Note: This list is based upon the timezone database version 2019.1.

as well as phpinfo()

![image](https://user-images.githubusercontent.com/1047696/58546517-cf8bbe00-8205-11e9-8be4-42c795448373.png)

However, the intl extension can use a different TZ DB version. My local webserver seems to confirm

![image](https://user-images.githubusercontent.com/1047696/58546592-f9dd7b80-8205-11e9-99d7-4b5f2837fb7d.png)

This version already returns `Etc/Unknown` for `Europe/Saratov`. What version are you using?

ro0NL commented:
```bash
$ docker run -it jakzal/php-intl:7.2-49.1.2 bash
root@603af4e29326:/# php -i | grep TZ
ICU TZData version => 2012c
root@603af4e29326:/# php -r "var_dump('Etc/Unknown' === IntlTimeZone::createTimeZone('Europe/Saratov')->getID(), INTL_ICU_VERSION);"
bool(true)
string(6) "49.1.2"
```

mmokhi commented:
> ```shell
> $ docker run -it jakzal/php-intl:7.2-49.1.2 bash
> root@603af4e29326:/# php -i | grep TZ
> ICU TZData version => 2012c
> root@603af4e29326:/# php -r "var_dump('Etc/Unknown' === IntlTimeZone::createTimeZone('Europe/Saratov')->getID(), INTL_ICU_VERSION);"
> bool(true)
> string(6) "49.1.2"
> ```

Same command gives me this result:

```
bool(false)
string(4) "57.1"
```

So I assume mine is `57.1` :smile:

mmokhi commented:
@ro0NL , Did you get a chance to try with `America/Bering` instead of `Saratov`?

ro0NL commented:
`America/Bering` is already removed in ICU. For these tests we need a timezone that's in [the list](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Intl/Resources/data/timezones/en.json), but unavailable using `\IntlTimeZone` (depending on the TZ DB version of the intl extension).

It's an edge case situation :)

> Same command gives me this result:
> So I assume mine is 57.1

You should check for `ICU TZData version` in phpinfo on _your_ environment; the one where tests are failing and lead to this issue.

However in #31685 i figured `Europe/Saratov` can be used until TZ DB version 2017.

nicolas-grekas commented:
Replaced by #31685, thanks @mmokhi

mmokhi commented:
Thank you all :blush:
